README.md
Kill-Bill
=========
*Web Application Server*


Installation
------------

Using rebar3:

```erlang
{deps, [
	{kill_bill, {git, "https://github.com/jjmrocha/kill-bill.git", {branch, "master"}}}
]}.
```


Start Kill-Bill
---------------

Kill-Bill depends on cowboy, jsondoc, erlydtl and gibreel, you need to start cowboy and gibreel before kill-bill.

```erlang
%% Start cowboy
ok = application:start(crypto),
ok = application:start(ranch),
ok = application:start(cowlib),
ok = application:start(cowboy),
%% Start gibreel
ok = application:start(columbo),
ok = application:start(cclock),
ok = application:start(gibreel),
%% Start Kill-Bill
ok = application:start(kill_bill).
```


Server Management
-----------------

### API

```erlang
%% Setup a server
kill_bill:config_server(ServerConfig :: Config) -> {ok, ServerName :: atom()} | {error, Reason :: any()} 
	when Config :: ConfigFile :: string()
                     | {server_config, ServerName :: atom(), Config :: list()}.

%% Start server
kill_bill:start_server(ServerName :: atom()) -> ok | {error, Reason :: any()}.

%% Stop server
kill_bill:stop_server(ServerName :: atom()) -> ok | {error, Reason :: any()}.

%% List all servers
kill_bill:get_server_list() -> list().
```

### Server configuration file

```erlang
{server_config, ServerName, [
	{host, Host}, 
	{protocol, Protocol}, 
	{ssl, SSLConfig},
	{port, Port},
	{acceptor_number, AcceptorNumber},
	{max_connections, MaxConnections}
]}.
```

Where:

* ServerName - **Server name (it's an atom and is mandatory)**
* Host - **Host (defaults to '_' - all hosts), examples: "127.0.0.1", "www.gmailbox.org" or "[www.]gmailbox.org"**
* Protocol - **Http protocol (http or https - defaults to http)**
* SSLConfig - **Use with protocol = https (defaults to none), it's a proplist with the following properties: cacertfile, certfile and keyfile**
* Port - **Port number (defaults to 8080)**
* AcceptorNumber - **Number of processes accepting connections (defaults to 100)**
* MaxConnections - **Max number of connections (defaults to infinity)**

Example:
Accept connections on port 8080 for all hosts.
```erlang
{server_config, my_server, []}.
```


Web Application Management
--------------------------

### API

```erlang
%% Deploy a WebApp
kill_bill:deploy(ServerName :: atom(), WebAppConfig :: Config) -> ok | {error, Reason :: any()} 
	when Config :: WebAppFile :: string()
                     | {webapp_config, WebAppName :: atom(), Config :: list()}.	

%% Undeploy WebApp
kill_bill:undeploy(WebAppName :: atom()) -> ok | {error, Reason :: any()}.

%% List all WebApps
kill_bill:get_webapp_list() -> WebAppList
	when WebAppList :: [{WebAppName :: atom(), ServerName :: atom()}, ...].
```

### WebApp configuration file

```erlang
{webapp_config, WebAppName, [
	{context, Context},
	{template, TemplateConfig},
	{resource, ResourceConfig},
	{action, ActionList},
	{static, StaticConfig},
	{session_timeout, SesionTimeout}
]}.
```

Where:

* WebAppName - **WebApp name (it's an atom and is mandatory)**
* Context - **WebApp context (defaults to "/")**
* TemplateConfig - **Erlydtl template access configuration (defaults to none), it's a proplist with the following properties: top_page and prefix**
	* top_page - **Default page (equivalent to index.html, defaults to "index")**
	* prefix - **URL to template, i.e. /Context/Prefix/TemplateName**
* ResourceConfig - **Message files (defaults to none), it's a proplist with the following properties: base_name, file_extension and file_dir**
	* base_name - **Name of the message file (defaults to "message"), the full format is: file_dir/base_name[_language[_country]]file_extension**
	* file_extension - **Extension of the message files (defaults to ".txt")**
	* file_dir - **Directory where the message files are located (defaults to "./resource")**
* ActionList - **List of actions (prefix and callback pairs), each action is define as: {ActionPrefix, ActionCallback}, where:**
	* ActionPrefix - **URL of action, i.e. /Context/ActionPrefix**
	* ActionCallback - **Name of the erlang module, must implement the ```kb_action_handler``` callback behaviour**
* StaticConfig - **Static content configuration (defaults to none), it's a proplist with the following properties: path and dir**
	* path - **URL to file (defaults to "/"), i.e. /Context/Path/FileName**
	* dir - **Directory with static content (defaults to "./static")**
* SesionTimeout - **Session timeout in minutes (defaults to 30)**

### Example

The following code was removed from the example project, that you will find on the "examples" directory.
```erlang
{ok, ServerID} = kill_bill:config_server("./priv/default-server.config"),
ok = kill_bill:deploy(ServerID, "./priv/root-webapp.config"),
ok = kill_bill:deploy(ServerID, "./priv/examples-webapp.config"),
ok = kill_bill:start_server(ServerID).
```


kb_action_handler callback
--------------------------

The actions must implement the ```kb_action_handler``` callback behaviour.

```erlang
handle(Method :: binary(), Path :: list(), Request :: #kb_request{}) 
	-> {html, Value :: iodata(), Request1 :: #kb_request{}} 
	 | {json, Value :: any(), Request1 :: #kb_request{}}
	 | {dtl, Template :: module(), Args :: list(), Request1 :: #kb_request{}}
	 | {redirect, Url :: iolist(), Request1 :: #kb_request{}} 
	 | {raw, Status :: integer(), Headers :: list(), Body :: iodata(), Request1 :: #kb_request{}}.
```

Parameters:

* Method - **HTTP Method (GET, POST, ...)**
* Path - **The URL split by "/", i.e. the request ```GET /shop/hardware``` will produce the path = [<<"shop">>, <<"hardware">>]**
* Request - **Record with the request data**

### Example

WebApp configuration file:
```erlang
{webapp_config, examples, [
	{context, "/examples/"},
	{action, [ 
		{"/", examples_action}		
	]}
]}.
```

Action module:
```erlang
-module(examples_action).

-behaviour(kb_action_handler).

-export([handle/3]).

handle(_Method, [], Req) ->
	{html, "<html><body>Kill-Bill examples.</body></html>", Req};

handle(_Method, [<<"obj">>, Id, <<"create">>], Req) ->
	Obj = [{id, Id}, {type, <<"obj">>}],
	{json, Obj, Req};

handle(<<"GET">>, [<<"cities">>], Req) ->
	CitiesList = [<<"Lisbon">>, <<"London">>, <<"Madrid">>],
	Args = [{cities, ValueList}],
	{dtl, cities_dtl, Args, Req};

handle(_Method, [<<"google">>], Req) ->
	{redirect, <<"http://www.google.com">>, Req};

handle(_Method, _Path, Req) ->
	{raw, 404, [], "Not found!", Req}.
```


kb_action_helper Helper module
------------------------------

The module ```kb_action_helper``` provides functions to interact with Kill-Bill.

### API

```erlang
%% Retrive WebApp context
kb_action_helper:get_context(Req :: #kb_request{}) -> binary().

%% Retrive args from request
kb_action_helper:get_args(Req :: #kb_request{}) -> {Args, Req1 :: #kb_request{}}
	when Args :: [{Key :: binary(), Value :: binary()}, ...].

%% Retrive JSON from request body
kb_action_helper:get_json(Req :: #kb_request{}) -> {Doc :: jsondoc:jsondoc(), Req1 :: #kb_request{}}.

%% Retrive data from session
kb_action_helper:get_session(Req :: #kb_request{}) -> {SessionData, Req1 :: #kb_request{}}
	when SessionData :: [{Key :: any(), Value :: any()}, ...].

%% Store data in session
kb_action_helper:set_session(SessionData, Req :: #kb_request{}) -> Req1 :: #kb_request{}
	when SessionData :: [{Key :: any(), Value :: any()}, ...].

%% Invalidate session
kb_action_helper:invalidate_session(Req :: #kb_request{}) -> #kb_request{}.
```

*The module ```kb_action_helper``` has many more functions.*


Kill-Bill tags
--------------

Kill-Bill provides two tags to be use on templates:

### TAG context_path

Without parameters, returns the WebApp's context
```
<form action="{% context_path %}action" method="post">
...
</form>
```

With the "file" parameter, the TAG prefixes the path to the file with the WebApp's context:
```
<img src="{% context_path file="/images/kill-bill-logo.png" %}">
```

### TAG message

Retrieves the localized message for "key" from the messages file (defined in the WebApp configuration file):
```
<h1>{% message key="page_title" %}</h1>
```

### Configuration

To use Kill-Bill's tags, add the following to your project's rebar.config:

```erlang
{erlydtl_opts, [
	{libraries, [{kb_dtl_tag, kb_dtl_tag}]},
	{default_libraries, [kb_dtl_tag]}
]}.
```

To compile a DTL directory with rebar3 compile add the following to your project's rebar.config:
```erlang
{provider_hooks, [
	{post, [{compile, {erlydtl, compile}}]}
].

{plugins, [
	{rebar3_erlydtl_plugin, ".*", {git, "https://github.com/tsloughter/rebar3_erlydtl_plugin.git", {branch, "master"}}}
]}.
```

CHANGELOG.md
# Changelog

## [1.2.0] - 2018-06-28

- Log requests for debugging purposes

### Changed

```erlang
- kb_action_helper.erl - log the request body after reading.
- kb_cowboy_action:init(...) - check the flag before logging and add an unique id to the log string.
- kb_util.erl - new log function.
- kill_bill.hrl - new fields in kb_request record.
- kill_bill_sup:init(...) - load the flag from applications' env if exists.
- kill_bill.erl - new function log_actions/2 to change the log flag.
```

## [1.1.0] - 2017-12-18

- Upgraded to rebar3.
- Upgraded cowboy (tag 2.2.0) and erlydtl (tag 0.12.1).

### Changed
- The following functions no longer return the changed request

```erlang
kb_action_helper:get_accept_header
kb_action_helper:get_content_type
kb_action_helper:get_cookie
kb_action_helper:get_headers
kb_http:get_cookie
kb_http:get_header
kb_locale:get_accept_languages
kb_session:get_session_id
kb_session:get_session
kb_session:touch_session
```

- Renamed

```erlang
kb_session:touch_session(CacheName :: atom(), SessionID :: binary(), Data :: cowboy_req:req())
```
to

```erlang
kb_session:touch_session_id(CacheName :: atom(), SessionID :: binary())
```
- Renamed custom tag kb_dtl_tag:context to context_path



src/kill_bill.erl
%%
%% Copyright 2013-16 Joaquim Rocha
%% 
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%   http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.
%%
%% Modifications copyright (C) 2017-18 Sysvision, Lda.
%%

-module(kill_bill).

-behaviour(gen_server).

-define(PROTOCOL_HTTP, http).
-define(PROTOCOL_HTTPS, https).

-define(SERVER, {local, ?MODULE}).

-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).
-export([start_link/1]).
-export([config_server/1, start_server/1, stop_server/1, get_server_list/0]).
-export([deploy/2, undeploy/1, get_webapp_list/0]).
-export([log_actions/2]).

%% ====================================================================
%% API functions
%% ====================================================================

start_link(Args) ->
	gen_server:start_link(?SERVER, ?MODULE, Args, []).

-spec config_server(ServerConfig :: Config) -> {ok, ServerName :: atom()} | {error, Reason :: any()} 
	when Config :: ConfigFile :: string()
	| {server_config, ServerName :: atom(), Config :: list()}.
config_server({server_config, ServerName, Config}) when is_atom(ServerName) andalso is_list(Config) ->
	gen_server:call(?MODULE, {config_server, ServerName, Config});
config_server(FileName) when is_list(FileName) ->
	case load_configuration(FileName) of
		not_found -> {error, not_found};
		{error, Reason} -> {error, Reason};
		{ok, Config} ->	config_server(Config)
	end.

-spec start_server(ServerName :: atom()) -> ok | {error, Reason :: any()}.
start_server(ServerName) when is_atom(ServerName) ->
	gen_server:call(?MODULE, {start_server, ServerName}).

-spec stop_server(ServerName :: atom()) -> ok | {error, Reason :: any()}.
stop_server(ServerName) when is_atom(ServerName) ->
	gen_server:call(?MODULE, {stop_server, ServerName}).

-spec get_server_list() -> list().
get_server_list() ->
	gen_server:call(?MODULE, {get_server_list}).

-spec deploy(ServerName :: atom(), WebAppConfig :: Config) -> ok | {error, Reason :: any()} 
	when Config :: WebAppFile :: string()
	| {webapp_config, WebAppName :: atom(), Config :: list()}.	  
deploy(ServerName, {webapp_config, WebAppName, Config}) when is_atom(ServerName) andalso is_atom(WebAppName) andalso is_list(Config) ->
	gen_server:call(?MODULE, {deploy, ServerName, {WebAppName, Config}});
deploy(ServerName, FileName) when is_atom(ServerName) andalso is_list(FileName) ->
	case load_configuration(FileName) of
		not_found -> {error, not_found};
		{error, Reason} -> {error, Reason};
		{ok, Config} -> deploy(ServerName, Config)
	end.

-spec undeploy(WebAppName :: atom()) -> ok | {error, Reason :: any()}.
undeploy(WebAppName) when is_atom(WebAppName) ->
	gen_server:call(?MODULE, {undeploy, WebAppName}).

-spec get_webapp_list() -> list().
get_webapp_list() ->
	gen_server:call(?MODULE, {get_webapp_list}).

-spec log_actions(WebAppName :: atom(), Flag :: boolean()) -> ok.
log_actions(WebAppName, Flag) when is_atom(WebAppName) andalso is_boolean(Flag) ->
	gen_server:call(?MODULE, {log_actions, WebAppName, Flag}).

%% ====================================================================
%% Behavioural functions 
%% ====================================================================

-record(server, {config, webapps = [], running=false}).
-record(webapp, {config, resource, server, session}).
-record(status, {servers, webapps, log_actions}).

init(LogActions) ->
	process_flag(trap_exit, true),
	error_logger:info_msg("Bang Bang, ~p [~p - LogActions: ~p] Starting...\n", [?MODULE, self(), LogActions]),
	{ok, #status{servers=dict:new(), webapps=dict:new(), log_actions=LogActions}}.

handle_call({config_server, ServerName, Config}, _From, State=#status{servers=Servers}) ->
	Reply = case dict:find(ServerName, Servers)  of
		error ->						
			NServers = dict:store(ServerName, #server{config=Config}, Servers),
			NState = State#status{servers=NServers},
			error_logger:info_msg("Server ~p configured: ~p\n", [ServerName, Config]),
			{ok, ServerName};
		{ok, #server{running=false, webapps=WebApps}} ->
			NServers = dict:store(ServerName, #server{config=Config, webapps=WebApps}, Servers),
			NState = State#status{servers=NServers},
			error_logger:info_msg("Server ~p reconfigured: ~p\n", [ServerName, Config]),
			{ok, ServerName};	
		{ok, _Server} ->
			error_logger:error_msg("KB: Server ~p is running\n", [ServerName]),
			NState = State,
			{error, server_running}
	end,
	{reply, Reply, NState};
handle_call({start_server, ServerName}, _From, State=#status{servers=Servers, webapps=Webapps, log_actions=LogActions}) ->
	Reply = case dict:find(ServerName, Servers)  of
		error ->			
			error_logger:error_msg("KB: Server ~p not found\n", [ServerName]),
			NState = State,
			{error, not_found};
		{ok, #server{running=true}} ->
			error_logger:info_msg("Server ~p was already running!\n", [ServerName]),
			NState = State,
			ok;
		{ok, Server} ->
			Protocol = proplists:get_value(protocol, Server#server.config, ?PROTOCOL_HTTP),
			Host = get_host(Server#server.config),
			Port = get_port(Server#server.config),
			
			PathsList = get_server_paths(ServerName, Host, Server, Webapps, LogActions),
			
			Dispatch = cowboy_router:compile([{Host, PathsList}]),
			ProtoOpts = #{env => #{dispatch => Dispatch}},
			
			TransOpts = get_server_config(Protocol, Port, Server#server.config),
			
			case Protocol of
				?PROTOCOL_HTTP -> 
					{ok, _} = cowboy:start_clear(ServerName, TransOpts, ProtoOpts);
				?PROTOCOL_HTTPS -> 
					{ok, _} = cowboy:start_tls(ServerName, TransOpts, ProtoOpts)
			end,
			
			error_logger:info_msg("Server ~p started!\n", [ServerName]),
			
			NServers = dict:store(ServerName, Server#server{running=true}, Servers),
			NState = State#status{servers=NServers},
			ok
	end,
	{reply, Reply, NState};
handle_call({stop_server, ServerName}, _From, State=#status{servers=Servers}) ->
	Reply = case dict:find(ServerName, Servers)  of
		error ->			
			error_logger:error_msg("KB: Server ~p not found\n", [ServerName]),
			NState = State,
			{error, not_found};
		{ok, #server{running=false}} ->
			error_logger:info_msg("Server ~p was not running!\n", [ServerName]),
			NState = State,
			ok;
		{ok, Server} ->
			cowboy:stop_listener(ServerName),
			error_logger:info_msg("Server ~p stoped!\n", [ServerName]),
			
			NServers = dict:store(ServerName, Server#server{running=false}, Servers),
			NState = State#status{servers=NServers},
			ok
	end,
	{reply, Reply, NState};
handle_call({get_server_list}, From, State=#status{servers=Servers}) ->
	Fun = fun() ->
			Rep = fun(ServerName, #server{running=Run, webapps=WebApps}, Acc) ->
					[{ServerName, Run, WebApps} | Acc]
			end,
			Reply = dict:fold(Rep, [], Servers),
			gen_server:reply(From, Reply)
	end,
	spawn(Fun),
	{noreply, State};
handle_call({deploy, ServerName, {WebAppName, Config}}, _From, State=#status{servers=Servers, webapps=Webapps, log_actions=LogActions}) ->
	Reply = case dict:find(WebAppName, Webapps) of
		error ->
			case dict:find(ServerName, Servers) of
				error ->
					error_logger:error_msg("KB: Server ~p not configured\n", [ServerName]),
					NState = State,
					{error, not_found};
				{ok, Server} ->
					Resource = create_resource_server(Config),
					SessionManager = create_session(WebAppName, Config),					
					WebApp = #webapp{config=Config, resource=Resource, server=ServerName, session=SessionManager},
					NWebapps = dict:store(WebAppName, WebApp, Webapps),
					
					NServer = Server#server{webapps=[WebAppName | Server#server.webapps]},
					NServers = dict:store(ServerName, NServer, Servers),
					
					case NServer#server.running of
						true ->
							Host = get_host(NServer#server.config),
							PathsList = get_server_paths(ServerName, Host, NServer, NWebapps, LogActions),
							Dispatch = cowboy_router:compile([{Host, PathsList}]),
							cowboy:set_env(ServerName, dispatch, Dispatch);
						false -> ok
					end,
					
					error_logger:info_msg("WebApp ~p deployed on server ~p, with Config: ~p\n", [WebAppName, ServerName, Config]),
					NState = State#status{servers=NServers, webapps=NWebapps},
					ok
			end;
		{ok, _App} ->
			error_logger:error_msg("KB: Duplicated webapp ~p\n", [WebAppName]),
			NState = State,
			{error, duplicated}
	end,
	{reply, Reply, NState};
handle_call({undeploy, WebAppName}, _From, State=#status{servers=Servers, webapps=Webapps, log_actions=LogActions}) ->
	Reply = case dict:find(WebAppName, Webapps) of
		error ->			
			error_logger:error_msg("KB: WebApp ~p not found\n", [WebAppName]),
			NState = State,
			{error, not_found};
		{ok, WebApp} ->
			NWebapps = dict:erase(WebAppName, Webapps),
			{ok, Server} = dict:find(WebApp#webapp.server, Servers),
			NServer = Server#server{webapps=lists:delete(WebAppName, Server#server.webapps)},
			NServers = dict:store(WebApp#webapp.server, NServer, Servers),
			
			case NServer#server.running of
				true ->
					Host = get_host(NServer#server.config),
					PathsList = get_server_paths(WebApp#webapp.server, Host, NServer, NWebapps, LogActions),
					
					Dispatch = cowboy_router:compile([{Host, PathsList}]),
					cowboy:set_env(WebApp#webapp.server, dispatch, Dispatch);
				false -> ok
			end,
			
			stop_resource_server(WebApp#webapp.resource),
			stop_session(WebApp#webapp.session),
			
			error_logger:info_msg("WebApp ~p was undeployed from server ~p\n", [WebAppName, WebApp#webapp.server]),
			NState = State#status{servers=NServers, webapps=NWebapps},
			ok
	end,
	{reply, Reply, NState};
handle_call({get_webapp_list}, From, State=#status{webapps=Webapps}) ->
	Fun = fun() ->
			Rep = fun(WebAppName, #webapp{server=ServerName}, Acc) ->
					[{WebAppName, ServerName} | Acc]
			end,
			Reply = dict:fold(Rep, [], Webapps),
			gen_server:reply(From, Reply)
	end,
	spawn(Fun),
	{noreply, State};
handle_call({log_actions, WebAppName, Flag}, _From, State=#status{servers=Servers, webapps=Webapps}) ->
	Reply = case dict:find(WebAppName, Webapps) of
		error ->			
			error_logger:error_msg("KB: WebApp ~p not found\n", [WebAppName]),
			NState = State,
			{error, not_found};
		{ok, WebApp} ->
			{ok, Server} = dict:find(WebApp#webapp.server, Servers),
			case Server#server.running of
				true ->
					Host = get_host(Server#server.config),
					PathsList = get_server_paths(WebApp#webapp.server, Host, Server, Webapps, Flag),
					Dispatch = cowboy_router:compile([{Host, PathsList}]),
					cowboy:set_env(WebApp#webapp.server, dispatch, Dispatch);
				false -> ok
			end,
			error_logger:info_msg("WebApp ~p changed log_actions flag to ~p\n", [WebAppName, Flag]),
			NState = State#status{log_actions=Flag},
			ok
	end,
	{reply, Reply, NState}.

handle_cast(Msg, State) ->
	error_logger:info_msg("handle_cast(~p)\n", [Msg]),
	{noreply, State}.

handle_info(Info, State) ->
	error_logger:info_msg("handle_info(~p)\n", [Info]),
	{noreply, State}.

terminate(_Reason, _State) ->
	error_logger:info_msg("Bang Bang, My Baby Shot Me Down\n"),
	ok.

code_change(_OldVsn, State, _Extra) ->
	{ok, State}.


%% ====================================================================
%% Internal functions
%% ====================================================================

load_configuration(FileName) ->
	case filelib:is_file(FileName) of
		true ->	file:script(FileName);
		false -> not_found
	end.

get_host(ServerConfig) ->
	proplists:get_value(host, ServerConfig, '_').

get_port(ServerConfig) ->
	proplists:get_value(port, ServerConfig, 8080).

get_server_config(Protocol, Port, ServerConfig) ->
	SslConfig = proplists:get_value(ssl, ServerConfig),
	Ssl = get_ssl(Protocol, SslConfig),
	lists:append([{port, Port}], Ssl).

get_ssl(?PROTOCOL_HTTP, _Ssl) -> [];
get_ssl(_Protocol, none) -> [];
get_ssl(_Protocol, Ssl) -> 
	[
		{cacertfile, proplists:get_value(cacertfile, Ssl)},
		{certfile, proplists:get_value(certfile, Ssl)},
		{keyfile, proplists:get_value(keyfile, Ssl)}
		].

create_resource_server(WebAppConfig) ->
	ResourceConfig = proplists:get_value(resource, WebAppConfig, none),
	get_resource_server(ResourceConfig).

get_resource_server(none) -> none;
get_resource_server(ResourceConfig) ->
	{ok, Pid} = kb_resource_sup:start_resource_server(ResourceConfig),
	Pid.

stop_resource_server(none) -> ok;
stop_resource_server(Pid) -> kb_resource:stop(Pid).

create_session(WebAppName, WebAppConfig) ->
	case proplists:get_value(session_timeout, WebAppConfig, 30) of
		none -> none;
		SessionTimeout  ->
			SessionCache = list_to_atom(atom_to_list(WebAppName) ++ "_session"),
			Options = [{max_age, SessionTimeout * 60},
					{purge_interval, 60},
					{cluster_nodes, all},
					{sync_mode, full}],
			gibreel:create_cache(SessionCache, Options),
			SessionCache
	end.

stop_session(none) -> ok;
stop_session(SessionCache) ->
	gibreel:delete_cache(SessionCache).

get_server_paths(ServerName, Host, Server, Webapps, LogActions) ->
	case Server#server.webapps of 
		[] -> 
			[{'_', kb_dummy_toppage, [{server, ServerName}, {host, Host}]}];
		WAList -> 
			AppList = get_app_list(WAList, Webapps, []),
			get_web_app_config(AppList, [], LogActions)
	end.

get_app_list([], _Webapps, AppList) ->
	Fun = fun({_, A, _}, {_, B, _}) -> A > B end,
	lists:sort(Fun, AppList);
get_app_list([WebAppName|T], Webapps, AppList) ->
	{ok, WebApp} = dict:find(WebAppName, Webapps),
	Context = get_context(proplists:get_value(context, WebApp#webapp.config, "/")),
	get_app_list(T, Webapps, [{WebAppName, Context, WebApp} | AppList]).

remove_slashs(Path) ->
	kb_util:remove_if_ends_with(kb_util:remove_if_starts_with(Path, "/"), "/").

get_web_app_config([], Paths, _LogActions) -> lists:reverse(Paths);
get_web_app_config([{_WebAppName, Context, WebApp} | T], Paths, LogActions) ->
	ResourceServer = WebApp#webapp.resource,
	SessionManager = WebApp#webapp.session,
	TemplateConfig = proplists:get_value(template, WebApp#webapp.config, none),
	ActionConfig = proplists:get_value(action, WebApp#webapp.config, []),
	StaticConfig = proplists:get_value(static, WebApp#webapp.config, none),
	Static = get_static(StaticConfig),
	Options = [{resource_server, ResourceServer},
	           {context, Context},
	           {session_manager, SessionManager},
	           {static, Static},
	           {log_actions, LogActions}],
	ActionPath = action_path(ActionConfig, [], []),
	PathsWithTemplate = add_template(TemplateConfig, Context, Options, []),
	PathsWithAction = add_action(ActionPath, Context, Options, PathsWithTemplate),
	PathsWithStatic = add_static(StaticConfig, Context, PathsWithAction),
	Sorted = lists:sort(fun({A, _, _}, {B, _, _}) -> sort_paths(A, B) end, PathsWithStatic),
	get_web_app_config(T, lists:append(Sorted, Paths), LogActions).

get_static(none) -> "";
get_static(Config) ->
	Path = proplists:get_value(path, Config, "/"), 
	remove_slashs(Path) ++ "/".

action_path([{Filter, SubConfig}|T], Path, Output) when is_atom(Filter) andalso is_list(SubConfig) ->
	NewOutput = action_path(SubConfig, [Filter|Path], []),
	action_path(T, Path, Output ++ NewOutput);
action_path([{ActionContext, Action}|T], Path, Output) ->
	CallbackList = lists:reverse([Action|Path]),
	action_path(T, Path, [{ActionContext, CallbackList}|Output]);
action_path([], _Path, Output) -> Output.

sort_paths(A, B) ->
	{UA, QA} = fix_path(A),
	{UB, QB} = fix_path(B),
	if UA < UB -> true;
		UA == UB -> QA > QB;
		true -> false
	end.

fix_path(Path) ->
	case string:rstr(Path, "[...]") of
		0 -> {Path, ""};
		Len -> 
			Url = string:substr(Path, 1, Len - 1),
			{Url, "[...]"}
	end.

add_template(none, _Context, _Options, Paths) -> Paths;
add_template(TemplateConfig, Context, Options, Paths) ->
	TemplatePrefix = proplists:get_value(prefix, TemplateConfig, "page"),
	TopPage = proplists:get_value(top_page, TemplateConfig, "index"),
	NPaths = add_top_page(TopPage, Context, Options, Paths),
	lists:append([
			{get_template_match(TemplatePrefix, Context), kb_cowboy_template, Options}
			], NPaths).

add_top_page(none, _Context, _Options, Paths) -> Paths;
add_top_page(TopPage, Context, Options, Paths) -> 
	lists:append([
			{Context, kb_cowboy_toppage, [{top_page, TopPage}] ++ Options}
			], Paths).	

get_template_match(TemplatePrefix, Context) ->
	Context ++ remove_slashs(TemplatePrefix) ++ "/[...]".

add_action([], _Context, _Options, Paths) -> Paths;
add_action([{ActionPrefix, CallbackList}|T], Context, Options, Paths) ->
	NakedActionPrefix = remove_slashs(ActionPrefix),
	NPaths = lists:append([
				{get_action_match(NakedActionPrefix, Context), kb_cowboy_action, [{callback_list, CallbackList}, {action_prefix, NakedActionPrefix}] ++ Options}
				], Paths),
	add_action(T, Context, Options, NPaths).

get_action_match(ActionPrefix, Context) ->
	Context ++ ActionPrefix ++ "/[...]".

add_static(none, _Context, Paths) -> Paths;
add_static(StaticConfig, Context, Paths) ->
	Path = proplists:get_value(path, StaticConfig, "/"), 
	Options = [{mimetypes, cow_mimetypes, all}],
	Config = case lists:keyfind(priv_dir, 1, StaticConfig) of
		{_, App, Dir} -> {priv_dir, App, Dir, Options};
		false -> 
			Dir = proplists:get_value(dir, StaticConfig, "./static"),
			{dir, Dir, Options}
	end,	
	lists:append([
			{get_static_match(Path, Context), cowboy_static, Config}
			], Paths).

get_static_match(Path, Context) ->
	case Path of
		"/" -> string:concat(Context, "[...]");
		_ -> Context ++ remove_slashs(Path) ++ "/[...]"
	end.

get_context("/") -> "/";
get_context(WebContext) -> 
	"/" ++ remove_slashs(WebContext) ++ "/".